src/parser/sparqlParser/SparqlQleverVisitor.cpp

```diff
@@ -126,29 +126,21 @@ PathTuples joinPredicateAndObject(VarOrPath predicate, ObjectList objectList) {
 }
 
 // ____________________________________________________________________________________
-ParsedQuery Visitor::visitTypesafe(Parser::QueryContext* ctx) {
+ParsedQuery Visitor::visit(Parser::QueryContext* ctx) {
   // The prologue (BASE and PREFIX declarations)  only affects the internal
   // state of the visitor.
-  visitTypesafe(ctx->prologue());
-  ParsedQuery query;
-  // TODO<joka921, qup42> Check if there are more instances of this pattern
-  //  (like `visitAlternative`, but with a custom error message), that would
-  //  justify extracting this pattern.
-  if (ctx->selectQuery()) {
-    query = visitTypesafe(ctx->selectQuery());
-  } else if (ctx->constructQuery()) {
-    query = visitTypesafe(ctx->constructQuery());
-  } else {
-    reportError(ctx, "QLever only supports SELECT and CONSTRUCT queries.");
-  }
+  visit(ctx->prologue());
+  auto query =
+      visitAlternative<ParsedQuery>(ctx->selectQuery(), ctx->constructQuery(),
+                                    ctx->describeQuery(), ctx->askQuery());
```

Comment:
Where it made sense I decided to have the same return type (here `ParsedQuery`) for the unsupported branches (here `describeQuery` and `askQuery`) as the real branches. This allows to use `visitAlternative` here and moves the throwing of error into the corresponding functions. Although the return type is slightly misleading the functions are still annotated with `[[noreturn]]`. I'd currently say that this approach makes sense if there is a natural `fake return type` to use.

Reply:
I mean at least in this case it makes perfect sense that the ...query rules return a `ParsedQuery`, but I'll look at the rest.

Reply:
Ah I understand, you call the unsupported clauses and report the error there, and then you can use `visitAlternative` more often. I approve of this approach.

Reply:
It is not the case here but I also had another thing in mind. Sometimes unsupported clauses can be reached from different clauses. This would also us to remove duplicated throws. Currently we often have the throw before actually calling the unsupported clause.